mllib/src/test/scala/org/apache/spark/ml/clustering/KMeansSuite.scala

```diff
@@ -137,18 +143,64 @@ class KMeansSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultR
       assert(model.clusterCenters === model2.clusterCenters)
     }
     val kmeans = new KMeans()
-    testEstimatorAndModelReadWrite(kmeans, dataset, KMeansSuite.allParamSettings, checkModelData)
+    testEstimatorAndModelReadWrite(kmeans, dataset, KMeansSuite.allParamSettings, checkModelData,
+      Map("initialModel" -> (checkModelData _).asInstanceOf[(Any, Any) => Unit]))
+  }
+
+  test("Initialize using a trained model") {
+    val kmeans = new KMeans().setK(k).setSeed(1).setMaxIter(1)
+    val oneIterModel = kmeans.fit(dataset)
+    val twoIterModel = kmeans.copy(ParamMap(ParamPair(kmeans.maxIter, 2))).fit(dataset)
+    val oneMoreIterModel = kmeans.setInitialModel(oneIterModel).fit(dataset)
+
+    twoIterModel.clusterCenters.zip(oneMoreIterModel.clusterCenters)
+      .foreach { case (center1, center2) => assert(center1 ~== center2 absTol 1E-8) }
+  }
+
+  test("Initialize using a model with wrong dimension of cluster centers") {
+    val kmeans = new KMeans().setK(k).setSeed(1).setMaxIter(1)
+
+    val wrongDimModel = KMeansSuite.generateRandomKMeansModel(4, k)
+    val wrongDimModelThrown = intercept[IllegalArgumentException] {
+      kmeans.setInitialModel(wrongDimModel).fit(dataset)
+    }
+    assert(wrongDimModelThrown.getMessage.contains("mismatched dimension"))
+  }
+
+  test("Infer K from an initial model if K is unset") {
+    val kmeans = new KMeans()
+    val testNewK = 10
+    val randomModel = KMeansSuite.generateRandomKMeansModel(dim, testNewK)
+    assert(kmeans.setInitialModel(randomModel).getK === testNewK)
+  }
+
+  test("Initialize using a model with wrong K if K is set") {
+    val kmeans = new KMeans().setK(k).setSeed(1).setMaxIter(1)
+
+    val wrongKModel = KMeansSuite.generateRandomKMeansModel(3, k + 1)
+    val wrongKModelThrown = intercept[IllegalArgumentException] {
+      kmeans.setInitialModel(wrongKModel).fit(dataset)
```

Comment:
According to discussions below, `setInitialModel` only overwrites param k.

1. If we use `km.setInitialModel(m1)` or `km.setK(k).setInitialModel(m1)`, then km with `k=m1.getK`;

2. If we use `km.setInitialModel(m1).setInitialModel(m2)`, then km with `k = m2.getK`;

3. If we use `km.setInitialModel(m1).setK(differentK)`, then an exception will be thrown out when `km.fit(data)`.

I think the case 3 is a wrong use case, so it's reasonable to throw an exception. What do you think?